core/test/utils/fb_matrix_generator.hpp

```diff
@@ -0,0 +1,294 @@
+ * @param engine  a random engine
+ * @param exec  executor where the matrix should be allocated
+ * @param args  additional arguments for the matrix constructor
+ *
+ * @return the unique pointer to generated matrix of type MatrixType
+ */
+template <typename MatrixType = matrix::Dense<>, typename NonzeroDistribution,
+          typename ValueDistribution, typename Engine, typename... MatrixArgs>
+std::unique_ptr<MatrixType> generate_random_matrix_with_diag(
+    size_type num_rows, size_type num_cols, NonzeroDistribution &&nonzero_dist,
+    ValueDistribution &&value_dist, Engine &&engine,
+    std::shared_ptr<const Executor> exec, MatrixArgs &&... args)
+{
+    using value_type = typename MatrixType::value_type;
+    using index_type = typename MatrixType::index_type;
+
+    matrix_data<value_type, index_type> data{gko::dim<2>{num_rows, num_cols},
+                                             {}};
+
+    for (size_type row = 0; row < num_rows; ++row) {
+        std::vector<size_type> col_idx(num_cols);
+        std::iota(col_idx.begin(), col_idx.end(), size_type(0));
+        // randomly generate number of nonzeros in this row
+        auto nnz_in_row = static_cast<size_type>(nonzero_dist(engine));
+        nnz_in_row = std::max(size_type(1), std::min(nnz_in_row, num_cols));
+        // select a subset of `nnz_in_row` column indexes, and fill these
+        // locations with random values
+        std::shuffle(col_idx.begin(), col_idx.end(), engine);
+        // add diagonal if it does not exist
+        auto it = std::find(col_idx.begin(), col_idx.begin() + nnz_in_row, row);
+        if (it == col_idx.begin() + nnz_in_row) {
+            col_idx[nnz_in_row - 1] = row;
+        }
+        std::for_each(
+            begin(col_idx), begin(col_idx) + nnz_in_row, [&](size_type col) {
+                data.nonzeros.emplace_back(
+                    row, col,
+                    detail::get_rand_value<value_type>(value_dist, engine));
+            });
+    }
+
+    data.ensure_row_major_order();
+
+    // convert to the correct matrix type
+    auto result = MatrixType::create(exec, std::forward<MatrixArgs>(args)...);
+    result->read(data);
+    return result;
+}
+
+template <typename ValueType>
+inline std::enable_if_t<!gko::is_complex<ValueType>(), std::complex<ValueType>>
+complexify_if_possible(const std::complex<ValueType> x)
+{
+    using namespace std::complex_literals;
+    std::complex<ValueType> y{x};
+    constexpr ValueType eps = std::numeric_limits<ValueType>::epsilon();
+    constexpr ValueType minval = std::numeric_limits<ValueType>::min();
+    const ValueType absval = abs(x);
+    if (absval > minval && abs(y.imag / absval) < eps) y.imag = abs(x);
+    return y;
+}
+
+template <typename ValueType>
+inline std::enable_if_t<!gko::is_complex<ValueType>(), ValueType>
+complexify_if_possible(const ValueType x)
+{
+    return x;
+}
+
+
+/**
+ * Generates a block CSR matrix having the same sparsity pattern as
```

Comment:
the sparsity pattern is under block sense?

Reply:
Yes.